File: rap-extensions/bundles/org.eclipse.rap.rwt/js/rwt/remote/RemoteObjectFactory.js
/*******************************************************************************
 * Copyright (c) 2012, 2014 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/

(function(){

var ObjectRegistry = rwt.remote.ObjectRegistry;

rwt.remote.RemoteObjectFactory = { // TODO [tb] : merge with Connection.js? (not a factory)

  _db : {},

  getRemoteObject : function( target ) {
    var id = ObjectRegistry.getId( target );
    if( id == null ){
      throw new Error( "Invalid target for ServerObject, or target not in ObjectManager" );
    }
    return this._getRemoteObject( id );
  },

  remove : function( id ) {
    delete this._db[ id ];
  },

  _getRemoteObject : function( id ) {
    if( this._db[ id ] == null ) {
      this._db[ id ] = new rwt.remote.RemoteObject( id );
      this._initRemoteObject( id );
    }
    return this._db[ id ];
  },

  _initRemoteObject : function( id ) {
    var remoteObject = this._db[ id ];
    var handler = ObjectRegistry.getEntry( id ).handler;
    if( handler && handler.listeners ) {
      for( var i = 0; i < handler.listeners.length; i++ ) {
        var type = handler.listeners[ i ];
        remoteObject._.listen[ type ] = true;
      }
    }
  }

};


}());

File: rap-extensions/bundles/org.eclipse.rap.rwt/src/org/eclipse/rap/rwt/internal/protocol/RemoteObjectFactory.java
/*******************************************************************************
* Copyright (c) 2011, 2013 EclipseSource and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v1.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v10.html
*
* Contributors:
*    EclipseSource - initial API and implementation
*******************************************************************************/
package org.eclipse.rap.rwt.internal.protocol;

import org.eclipse.rap.rwt.internal.lifecycle.DisplayUtil;
import org.eclipse.rap.rwt.internal.remote.LifeCycleRemoteObject;
import org.eclipse.rap.rwt.internal.remote.RemoteObjectImpl;
import org.eclipse.rap.rwt.internal.remote.RemoteObjectRegistry;
import org.eclipse.rap.rwt.internal.util.ParamCheck;
import org.eclipse.rap.rwt.internal.lifecycle.WidgetUtil;
import org.eclipse.rap.rwt.remote.RemoteObject;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Widget;


/**
 * The methods of this class are used to create remote objects, only for use in LCAs.
 *
 * @see RemoteObject
 */
public final class RemoteObjectFactory {

  public static RemoteObject getRemoteObject( Widget widget ) {
    ParamCheck.notNull( widget, "widget" );
    return getForId( WidgetUtil.getId( widget ) );
  }

  public static RemoteObject getRemoteObject( Display display ) {
    ParamCheck.notNull( display, "display" );
    return getForId( DisplayUtil.getId( display ) );
  }

  public static RemoteObject getRemoteObject( String id ) {
    ParamCheck.notNull( id, "id" );
    return getForId( id );
  }

  public static RemoteObject createRemoteObject( Widget widget, String type ) {
    ParamCheck.notNull( widget, "widget" );
    ParamCheck.notNull( type, "type" );
    return createForId( WidgetUtil.getId( widget ), type );
  }

  public static RemoteObject createRemoteObject( String id, String type ) {
    ParamCheck.notNull( id, "id" );
    ParamCheck.notNull( type, "type" );
    return createForId( id, type );
  }

  private static RemoteObject createForId( String id, String type ) {
    LifeCycleRemoteObject remoteObject = new LifeCycleRemoteObject( id, type );
    RemoteObjectRegistry.getInstance().register( remoteObject );
    return remoteObject;
  }

  private static RemoteObject getForId( String id ) {
    RemoteObjectImpl remoteObject = RemoteObjectRegistry.getInstance().get( id );
    // TODO [rst] Required for LCA tests, remove lazy initialization
    if( remoteObject == null ) {
      remoteObject = new LifeCycleRemoteObject( id, null );
      RemoteObjectRegistry.getInstance().register( remoteObject );
    }
    return remoteObject;
  }

  private RemoteObjectFactory() {
    // prevent instantiation
  }

}
